linked_list/linked_list.py

```diff
@@ -86,43 +178,107 @@ def visit(self):
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        if self.__head == None: 
+            return None
+
+        if self.__head == self.__tail: 
+            return self.__head
+
+        current = self.__head
+
+        while current != None: 
+            prev = current.prev
+            next = current.next
+            current.next = prev
+            current.prev = next
+            current = next
+
+        temp = self.__head
+        self.__head = self.__tail
+        self.__tail = temp   
+            
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(log n)
+    # Space Complexity: O(1)
     def find_middle_value(self):
-        pass
+        if self.__head == None: 
+            return None
+
+        if self.__head == self.__tail: 
+            return self.__head
+
+        left = self.__head
+        left_index = 0
+        right = self.__tail
+        right_index = (self.length() - 1)
+
+        while left_index < right_index: 
+            left = left.next 
+            left_index += 1
+
+            right = right.prev
+            right_index -= 1
+
+        return left.value
+        
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_nth_from_end(self, n):

```

Comment:
👍  Good use of `length` and some arithmetic.   You could also have used the `get_at_index` method